polyfill/lib/ecmascript.mjs

```diff
@@ -3540,216 +3540,103 @@ export function UnbalanceDateDurationRelative(years, months, weeks, days, larges
+    (largestUnit !== 'year' && largestUnit !== 'month' && largestUnit !== 'week')
+  ) {
     return { years, months, weeks, days };
   }
 
   if (!plainRelativeTo) throw new RangeError(`a starting point is required for ${largestUnit}s balancing`);
 
-  const oneYear = new TemporalDuration(sign);
-  const oneMonth = new TemporalDuration(0, sign);
-  const oneWeek = new TemporalDuration(0, 0, sign);
-
-  // Perform arithmetic in the mathematical integer domain
-  years = bigInt(years);
-  months = bigInt(months);
-  weeks = bigInt(weeks);
-  days = bigInt(days);
+  const untilOptions = ObjectCreate(null);
+  untilOptions.largestUnit = largestUnit;
 
   switch (largestUnit) {
     case 'year': {
-      // balance days up to years
-      let newRelativeTo, oneYearDays;
-      ({ relativeTo: newRelativeTo, days: oneYearDays } = MoveRelativeDate(calendarRec, plainRelativeTo, oneYear));
-      while (days.abs().geq(MathAbs(oneYearDays))) {
-        days = days.subtract(oneYearDays);
-        years = years.add(sign);
-        plainRelativeTo = newRelativeTo;
-        ({ relativeTo: newRelativeTo, days: oneYearDays } = MoveRelativeDate(calendarRec, plainRelativeTo, oneYear));
-      }
-
-      // balance days up to months
-      let oneMonthDays;
-      ({ relativeTo: newRelativeTo, days: oneMonthDays } = MoveRelativeDate(calendarRec, plainRelativeTo, oneMonth));
-      while (days.abs().geq(MathAbs(oneMonthDays))) {
-        days = days.subtract(oneMonthDays);
-        months = months.add(sign);
-        plainRelativeTo = newRelativeTo;
-        ({ relativeTo: newRelativeTo, days: oneMonthDays } = MoveRelativeDate(calendarRec, plainRelativeTo, oneMonth));
-      }
-
-      // balance months up to years
-      newRelativeTo = CalendarDateAdd(calendarRec, plainRelativeTo, oneYear);
-      const untilOptions = ObjectCreate(null);
-      untilOptions.largestUnit = 'month';
-      let untilResult = CalendarDateUntil(calendarRec, plainRelativeTo, newRelativeTo, untilOptions);
-      let oneYearMonths = GetSlot(untilResult, MONTHS);
-      while (months.abs().geq(MathAbs(oneYearMonths))) {
-        months = months.subtract(oneYearMonths);
-        years = years.add(sign);
-        plainRelativeTo = newRelativeTo;
-        newRelativeTo = CalendarDateAdd(calendarRec, plainRelativeTo, oneYear);
-        const untilOptions = ObjectCreate(null);
-        untilOptions.largestUnit = 'month';
-        untilResult = CalendarDateUntil(calendarRec, plainRelativeTo, newRelativeTo, untilOptions);
-        oneYearMonths = GetSlot(untilResult, MONTHS);
-      }
-      break;
+      // balance months and days up to years
+      const later = AddDate(calendarRec, plainRelativeTo, new TemporalDuration(years, months, 0, days));
+      const untilResult = CalendarDateUntil(calendarRec, plainRelativeTo, later, untilOptions);
+      return {
+        years: GetSlot(untilResult, YEARS),
+        months: GetSlot(untilResult, MONTHS),
+        weeks,
```

Comment:
I'm glad you spotted this! It's written like this because it's guaranteed to be unchanged from its input. But looking back at it now, I'm not sure that is correct. Adding `const later = AddDate(calendarRec, plainRelativeTo, new TemporalDuration(years, months, 0, days));`  (a few lines above this) and keeping `weeks` unchanged seems iffy with respect to RFC 5545, but also seems to lead to undesired results:
```js
const duration = Temporal.Duration.from({ weeks: 5 });
duration.round({ largestUnit: 'months', relativeTo: '2023-11-14' })
  // => duration of 5 weeks
  // should be 1 month, 5 days?
```
The "5 weeks" result matches the previous behaviour of BalanceDateDurationRelative with the loops, but this PR already did correct some other pre-existing rounding bugs due to the loop not respecting the order in which RFC 5545 adds units, largest to smallest (for example, [in this test](https://github.com/tc39/test262/pull/3957/files/8b5c4020948462b484d470fbe18d4dc0a5d97432..090438b74262dd8a1d2ce61ed1bbd23021e48eb4#diff-60247d76ce922d75656a5e8e52bb46f101c051d6b2dee3d42c52cd762fdd8015)). So I'm thinking that you spotted a bug in the algorithm here. Does that sound about right to you?

Reply:
Yes, I think you're correct. 1 month, 5 days should be the correct answer. 

Another wrinkle: if I remember correctly, `weeks` is a special unit that shows up in balanced durations only if either `largestUnit` or `smallestUnit` is `weeks`. If neither, then balancing should always remove `weeks` completely. I don't have time today to look up the history, but you may want to review older meeting notes and/or issues and/or PRs to make sure my fuzzy memory is correct!

So the interesting case here isn't 5 weeks, which has a clear-cut answer, but rather is 2 weeks and 6 weeks. Will the remaining weeks be converted to days or left as weeks?  I don't remember how the "weeks must be largest or smallest unit" rule above handles the case where, in `round` only one of the units needs to be provided by the caller.  I'm hoping that our meeting or repo history will provide some insight!

So sorry I don't have time to track this down myself right now. Given that I'm probably the originator of the design here, you can blame me if it's bad!

Reply:
As discussed in the champions meeting, we'll merge this as is for now, and if necessary update it as part of Part 3 (#2612).